komga/src/flyway/kotlin/db/migration/sqlite/V20210117175346__trash_bin_part_2.kt

```diff
@@ -0,0 +1,62 @@
+package db.migration.sqlite
+
+import mu.KotlinLogging
+import net.jpountz.xxhash.XXHashFactory
+import org.flywaydb.core.api.migration.BaseJavaMigration
+import org.flywaydb.core.api.migration.Context
+import org.springframework.jdbc.core.JdbcTemplate
+import org.springframework.jdbc.core.ResultSetExtractor
+import org.springframework.jdbc.datasource.SingleConnectionDataSource
+import java.io.FileInputStream
+import java.net.URL
+import java.nio.file.Path
+import java.nio.file.Paths
+
+private val logger = KotlinLogging.logger {}
+
+class V202101091156000__trash_bin_part_2 : BaseJavaMigration() {
+  override fun migrate(context: Context) {
+    val jdbcTemplate = JdbcTemplate(SingleConnectionDataSource(context.connection, true))
+
+    val totalCount = jdbcTemplate.queryForObject("select count(*) from BOOK", Long::class.java)
+    var currentCount = 0
+
+    logger.info { "Calculating hash for $totalCount books" }
+    while (existsRecordWithoutHash(jdbcTemplate)) {
+      val values = jdbcTemplate.queryForList("select b.ID, b.URL from BOOK b where b.FILE_HASH == '' limit 500")
+      currentCount += values.size
+
+      val params = values.map {
+        val hash = getHash(Paths.get((URL(it["URL"] as String).toURI())))
+        arrayOf(hash, it["ID"])
+      }.toList()
+      jdbcTemplate.batchUpdate("UPDATE BOOK SET FILE_HASH = ? WHERE ID = ?", params)
+      logger.info { "Hash calculation progress: $currentCount/$totalCount" }
+    }
+  }
+
+  fun getHash(path: Path): String {
+    val seed = -0x68b84d74L
+    val file = path.toFile()
+    val bytesToSkip = java.lang.Long.highestOneBit(file.length() / 100)
+
+    with(FileInputStream(file)) {
+      val buf = ByteArray(4096)
+      val hash64 = XXHashFactory.fastestInstance().newStreamingHash64(seed)
```

Comment:
I have concerns about using XXhash64 on 32bit systems. Komga is used on various hardware, including older versions of raspberry pis that are in 32bits.

Reply:
True, xxHash64 is slower on 32 bit systems but it still works. Using 32 bit version instead of 64 shouldn't be a problem as well. It'd just decrease hash length and increase collision chance, though it'd still be pretty low, especially since there's also file size comparison.